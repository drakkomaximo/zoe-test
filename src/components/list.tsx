import React, { useEffect, useState, FC } from 'react'
import { getUser } from '../service/user.service'

export interface ListProps{
    id: number,
    name: string,
    avatar: string,
    income: number
}

const List:FC<{listRender: any}> = ({listRender}) => {

    const [lists, setLists] = useState<ListProps[] | []>([])
    const [count, setCount] = useState<number>(3)
    const [storeNames, setstoreNames] = useState<string[] | []>([])

    const storeName = (name: string) =>{
        const filterNmae  =  lists.filter(
            (fname: any) => fname.name !== name
        )
        setLists(filterNmae)
        setstoreNames([...storeNames, name])
    }

    useEffect(() => {
        if(listRender.length > 3){
            const newlist = listRender.slice(0, count)
            console.log(newlist)
            setLists(newlist)
        }else{
            setLists(listRender)
        }
      }, [listRender, count])

    return (
        <div>
            {lists.map( (listsMap) => {
                return (
                    <p key={listsMap.id}>
                        <button onClick={()=>{storeName(listsMap.name)}}>
                        {listsMap.name}
                        </button>
                    </p>
                )
            })}
            {
                lists.length > 0 && (
                    <>
                    <button onClick={()=>{setCount( count + 3)}}>see more</button>
                    <button onClick={()=>{setCount( count - 3 < 3 ? 3 : count - 3)}}>see less</button>
                    </>
                )
            }
        </div>
    )
}

export default List
