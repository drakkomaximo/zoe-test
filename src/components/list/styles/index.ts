import styled from 'styled-components'

export const PeopleContainerInformation = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    margin-top: 2rem;
`

export const PeopleContainerInformationText = styled.p<{
    fSize: string
    fWeight: number
    colorText: string 
    textAlign: string
    marginText: string
    paddingText: string
}>`
    font-size: ${({fSize}): string=> fSize};
    font-weight: ${({fWeight}): number=> fWeight};
    color: ${({colorText}): string=> colorText};
    font-family: Helvetica;
    text-align: ${({textAlign}): string=> textAlign};
    margin: ${({marginText}): string=> marginText};
    padding: ${({paddingText}): string=> paddingText};
`

export const SearchContainer = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    width: 100%;
    max-width: 375px;
    padding: 0rem 2rem;
`

export const ButtonsContainer = styled.div`
    display: flex;
    justify-content: space-between;
`