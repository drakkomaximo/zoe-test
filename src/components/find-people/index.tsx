import { FC, useState } from 'react'
import { CashIcon } from '../../assets/icons/cash.icon'
import { LeftArrowIcon } from '../../assets/icons/left-arrow.icon'
import MatchButton from '../../common/buttons/match.button'
import SearchInput from '../../common/inputs/search.input'
import { getUser } from '../../service/user.service'
import { ListProps } from '../list'
import { SearchContainer } from './styles'
import PeopleInformation from './ui/people-information'

export const FindPeople:FC = () => {
    const [valueSearched, setValueSearched] = useState<number | ''>('')
  const [listSearched, setlistSearched] = useState<any>([])
  const validate = (n: number) =>{
    if(n < 99999){
      setValueSearched(n)
    }
  }
  const search = async () =>{
    if(valueSearched !== ''){
      if(!isNaN(valueSearched)){
        const response = await getUser()
        const filterIncomePositive = response.filter(
          (fIncome: ListProps) => (fIncome.income <= valueSearched + 10000) && (fIncome.income >= valueSearched - 10000)
        )
        setlistSearched(filterIncomePositive)
      }
    }else{
      setValueSearched('')
      setlistSearched('')
    }
    setValueSearched('')
    console.log(listSearched)
  }
    return (
        <>
        <PeopleInformation/>
        <SearchContainer>
          <SearchInput
            customIcon={<CashIcon />}
            typeInput={'number'}
            maxValuesLength={5}
            onChanceValues={(e: any): void=>{
              if(e.target.value !== ''){
                validate(parseInt(e.target.value))
              }else{
                setValueSearched('')
              }
              }}
            inputValue={valueSearched}
          />
        </SearchContainer>
        <MatchButton 
          action={search}
          title={'Get matches'}
          customIcon={<LeftArrowIcon style={{marginLeft: '4rem', marginRight: '1rem'}} />}
          />
        </>
    )
}
