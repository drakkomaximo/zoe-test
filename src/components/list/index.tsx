import { useEffect, useState, FC, useContext } from 'react'
import { useHistory } from 'react-router-dom'
import ShowButton from '../../common/buttons/show.button'
import UserCard from '../../common/cards/user.card'
import SearchSelect from '../../common/selects/search.select'
import { ListContext } from '../../context/ListContext'
import { ButtonsContainer } from './styles'
import ListInformation from './ui/list-information'

export interface ListProps{
    id: number,
    name: string,
    avatar: string,
    income: number
}

export const List:FC = () => {

    const history = useHistory()
    const { listSearched, valueSearched, count, setCount, addrtype } = useContext(ListContext)
    const [lists, setLists] = useState<ListProps[] | []>([])
    const [storeNames, setstoreNames] = useState<ListProps[] | []>([])

    const storeName = (data: ListProps) =>{
        const filterNmae  =  lists.filter(
            (fname: any) => fname.name !== data.name
        )
        setLists(filterNmae)
        setstoreNames([...storeNames, data])
    }

    const sortList = (type: string) =>{
        switch (type) {
            case 'id':
                lists.sort((a: any, b:any) => {
                    if (a.id > b.id) {
                      return 1;
                    }
                    if (a.id < b.id) {
                      return -1;
                    }
                    // a must be equal to b
                    return 0;
                  })
                break;
            case 'name':
                lists.sort((a: any, b:any) => {
                    if (a.name > b.name) {
                      return 1;
                    }
                    if (a.name < b.name) {
                      return -1;
                    }
                    // a must be equal to b
                    return 0;
                  })
                break;
            case 'high':
                lists.sort((a: any, b:any) => {
                    if (a.income > b.income) {
                      return 1;
                    }
                    if (a.income < b.income) {
                      return -1;
                    }
                    // a must be equal to b
                    return 0;
                  })
                break;
            case 'less':
                lists.sort((a: any, b:any) => {
                    if (a.income < b.income) {
                      return 1;
                    }
                    if (a.income > b.income) {
                      return -1;
                    }
                    // a must be equal to b
                    return 0;
                  })
                break;
        
            default:
                break;
        }
    }

    useEffect(() => {
        if(listSearched.length === 0 && valueSearched === ''){
            history.push('/')
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
      }, [])

    useEffect(() => {
        if(addrtype !== ''){
            sortList(addrtype)
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
      }, [addrtype])

    useEffect(() => {
        if(listSearched.length > 3){
            const newlist = listSearched.slice(0, count)
            console.log(newlist)
            setLists(newlist)
        }else{
            setLists(listSearched)
        }
      }, [listSearched, count])

    return (
        <div>
            <ListInformation value={valueSearched}/>
            <SearchSelect />
            <br />
            {lists.map( (listsMap) => {
                return (
                    <UserCard 
                    action={()=> {return storeName(listsMap)}}
                    key={listsMap.id}
                    avatar={listsMap.avatar}
                    name={listsMap.name}
                    income={listsMap.income}
                    id={listsMap.id}
                    
                    />
                )
            })}
            {
                lists.length > 0 && (
                    <ButtonsContainer>
                    <ShowButton 
                        action={()=>{setCount( count === 3 ? 3 : count - 3)}}
                        title={'Show less -'}
                        type={'LESS'}
                    />
                    <ShowButton 
                        action={()=>{setCount( count >= listSearched.length ? count : count + 3 )}}
                        title={'Show more +'}
                        type={'MORE'}
                    />
                    </ButtonsContainer>
                )
            }
        </div>
    )
}
