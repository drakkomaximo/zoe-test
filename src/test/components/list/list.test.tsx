import React from 'react';
import { shallow, mount } from 'enzyme';
import { List } from "../../../components/list";
import { ListContext } from '../../../context/ListContext';

describe('Test on <List />', () => {

    test(
        `It should render the component with the error message correctly: 
            1) if the context value (listSearched) is equals an empty array
            2) if the context value (valueSearched) is equals a number`, 
        () => {

        // expecifict context initial values
        const listSearched: any = []
        const valueSearched: number = 10000
        const wrapper = mount( 
            <ListContext.Provider value={
                {
                    listSearched,
                    valueSearched
                }}>
                <List />
            </ListContext.Provider>
            
            )
        expect( wrapper ).toMatchSnapshot()
    })

    // global context initial values
    const listSearched: any = [{"id":33,"name":"Rand","avatar":"https://randomuser.me/api/portraits/men/80.jpg","income":12580},{"id":34,"name":"Kathy","avatar":"https://randomuser.me/api/portraits/women/25.jpg","income":10604},{"id":35,"name":"Susan","avatar":"https://randomuser.me/api/portraits/women/21.jpg","income":13601},{"id":36,"name":"Robin","avatar":"https://randomuser.me/api/portraits/men/37.jpg","income":10021},{"id":37,"name":"Peter","avatar":"https://randomuser.me/api/portraits/men/49.jpg","income":12550},{"id":38,"name":"Diana","avatar":"https://randomuser.me/api/portraits/women/70.jpg","income":10603},{"id":39,"name":"Richard","avatar":"https://randomuser.me/api/portraits/lego/6.jpg","income":12018}]
    const valueSearched: number = 20000
    const addrtype: string = 'Select...'
    let count: number = 3

    test(
        `It should render the component correctly Part1: 
            1) if the context value (listSearched) have values
            2) if the context value (valueSearched) is equals a number
            3) if the context value (addrtype) is equals a Select...
            4) if the context value (count) is equals a 3`, 
        () => {

        
        const wrapper = mount( 
            <ListContext.Provider value={
                {
                    listSearched,
                    valueSearched,
                    addrtype,
                    count
                }}>
                <List />
            </ListContext.Provider>
            
            )
        expect( wrapper ).toMatchSnapshot()
    })

    test(
        `It should render the component correctly Part2: 
            1) if the context value (listSearched) have values
            2) if the context value (valueSearched) is equals a number
            3) if the context value (addrtype) is equals a Select...
            4) if the context value (count) is equals a 3
            5) if the user click 1 times the botton (see more), The list of Agents increment 3 
            5) if the user click 1 times the botton (see less), The list of Agents decrement 3`, 
        () => {
 
        const setCount = jest.fn()
        const setAddrType = jest.fn()
        const setInternalLabel = jest.fn()
        const wrapper = mount( 
            <ListContext.Provider value={
                {
                    listSearched,
                    valueSearched,
                    addrtype,
                    count,
                    setCount,
                    setAddrType,
                    setInternalLabel
                }}>
                <List />
            </ListContext.Provider>
            
            )

        const seeMore = wrapper.find('button').at(1)
        seeMore.simulate('click')
        
        expect( setCount ).toHaveBeenCalledTimes(1)
        expect( setAddrType ).toHaveBeenCalledTimes(1)
        expect( setInternalLabel ).toHaveBeenCalledTimes(1)
        expect( setCount ).toHaveBeenCalledWith(6)
        expect( setAddrType ).toHaveBeenCalledWith('Select...')
        expect( setInternalLabel ).toHaveBeenCalledWith('Select...')

        const seeLess = wrapper.find('button').at(0)
        seeLess.simulate('click')

        expect( setCount ).toHaveBeenCalledWith(3)
        expect( setAddrType ).toHaveBeenCalledWith('Select...')
        expect( setInternalLabel ).toHaveBeenCalledWith('Select...')

    })

})
