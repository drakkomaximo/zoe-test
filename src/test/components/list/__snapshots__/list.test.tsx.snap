// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <List /> It should render the component correctly Part1: 
            1) if the context value (listSearched) have values
            2) if the context value (valueSearched) is equals a number
            3) if the context value (addrtype) is equals a Select...
            4) if the context value (count) is equals a 3 1`] = `
<div
  className="sc-fKgJPI hdQWmr"
>
  <div
    className="sc-bkbkJK jSWBvy"
  >
    <p
      className="sc-iemWCZ eMmknN"
    >
      Your matches
    </p>
    <p
      className="sc-iemWCZ jNSNbF"
    >
      Your income: 
      <span
        className="sc-pNWdM fyKwqX"
      >
        $
        20,000
      </span>
    </p>
  </div>
  <div
    className="sc-crzoAE cwQYcI"
  >
    <p
      className="sc-dlnjwi ftNYDk"
    >
      Order agents by.
    </p>
    <div
      className="sc-ksluID bMLISG"
    >
      <div
        className="sc-fnVZcZ eVFifn"
        onClick={[Function]}
      >
        <p
          className="sc-dlnjwi dgTBpW"
        />
        <svg
          fill="none"
          height={7}
          viewBox="0 0 12 7"
          width={12}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1.415 0L6.00001 4.32839L10.585 0L12 1.3358L6.00001 7L0 1.3358L1.415 0Z"
            fill="#8B8B8B"
          />
        </svg>
      </div>
    </div>
  </div>
  <br />
  <div
    className="sc-kfYoZR cpxgYi"
  >
    <div
      className="sc-kEqXSa fyPaxq"
      onClick={[Function]}
    >
      <img
        alt="image"
        height="78px"
        src="https://randomuser.me/api/portraits/men/80.jpg"
        style={
          Object {
            "backgroundColor": "transparent",
            "border": "none",
            "borderRadius": "50%",
            "boxShadow": "none",
            "filter": "none",
            "margin": "0px",
            "objectFit": "cover",
            "opacity": "1",
            "padding": "0px",
            "transition": "0.25s",
          }
        }
        width="78px"
      />
      <div>
        <p
          className="sc-dlnjwi bkUepo"
        >
          Rand
        </p>
        <p
          className="sc-dlnjwi cbrSPi"
        >
          ID: 
          33
        </p>
        <p
          className="sc-dlnjwi krtMok"
        >
          Income 
          <span
            className="sc-pNWdM fyKwqX"
          >
            $
            12,580
          </span>
        </p>
      </div>
    </div>
    <div
      className="sc-kEqXSa fyPaxq"
      onClick={[Function]}
    >
      <img
        alt="image"
        height="78px"
        src="https://randomuser.me/api/portraits/women/25.jpg"
        style={
          Object {
            "backgroundColor": "transparent",
            "border": "none",
            "borderRadius": "50%",
            "boxShadow": "none",
            "filter": "none",
            "margin": "0px",
            "objectFit": "cover",
            "opacity": "1",
            "padding": "0px",
            "transition": "0.25s",
          }
        }
        width="78px"
      />
      <div>
        <p
          className="sc-dlnjwi bkUepo"
        >
          Kathy
        </p>
        <p
          className="sc-dlnjwi cbrSPi"
        >
          ID: 
          34
        </p>
        <p
          className="sc-dlnjwi krtMok"
        >
          Income 
          <span
            className="sc-pNWdM fyKwqX"
          >
            $
            10,604
          </span>
        </p>
      </div>
    </div>
    <div
      className="sc-kEqXSa fyPaxq"
      onClick={[Function]}
    >
      <img
        alt="image"
        height="78px"
        src="https://randomuser.me/api/portraits/women/21.jpg"
        style={
          Object {
            "backgroundColor": "transparent",
            "border": "none",
            "borderRadius": "50%",
            "boxShadow": "none",
            "filter": "none",
            "margin": "0px",
            "objectFit": "cover",
            "opacity": "1",
            "padding": "0px",
            "transition": "0.25s",
          }
        }
        width="78px"
      />
      <div>
        <p
          className="sc-dlnjwi bkUepo"
        >
          Susan
        </p>
        <p
          className="sc-dlnjwi cbrSPi"
        >
          ID: 
          35
        </p>
        <p
          className="sc-dlnjwi krtMok"
        >
          Income 
          <span
            className="sc-pNWdM fyKwqX"
          >
            $
            13,601
          </span>
        </p>
      </div>
    </div>
  </div>
  <div
    className="sc-hHEiqL bkWmxg"
  >
    <button
      className="sc-gKAaRy hbNDUu"
      onClick={[Function]}
    >
      <p
        className="sc-dlnjwi gSthXn"
      >
        Show less -
      </p>
    </button>
    <button
      className="sc-gKAaRy hbNDUu"
      onClick={[Function]}
    >
      <p
        className="sc-dlnjwi cYXUzV"
      >
        Show more +
      </p>
    </button>
  </div>
</div>
`;

exports[`Test on <List /> It should render the component with the error message correctly: 
            1) if the context value (listSearched) is equals an empty array
            2) if the context value (valueSearched) is equals a number 1`] = `
<div
  className="sc-fKgJPI hdQWmr"
>
  <div
    className="sc-bkbkJK jSWBvy"
  >
    <p
      className="sc-iemWCZ eMmknN"
    >
      Your matches
    </p>
    <p
      className="sc-iemWCZ jNSNbF"
    >
      Your income: 
      <span
        className="sc-pNWdM fyKwqX"
      >
        $
        10,000
      </span>
    </p>
  </div>
  <p
    className="sc-dlnjwi jgAYg"
  >
    No available Agents based on your income. Please try a different income value.
  </p>
  <p
    className="sc-dlnjwi dtVbnI"
    cursor="pointer"
    onClick={[Function]}
  >
    Click here to return.
  </p>
</div>
`;
